### ISSUE TITLE
CGAL::Surface_mesh_shortest_path assertion violation

### ISSUE THREAD

**Issue by clessig** [CGAL/cgal#4327]
_Please use the following template to help us solving your issue._

## Issue Details

_Describe your issue. Please be specific (compilation error, runtime error, unexpected behavior, wrong results, etc.)._

CGAL error: assertion violation!
Expression : _idx < data_.size()
File       : /usr/local/include/CGAL/Surface_mesh/Properties.h
Line       : 204
Explanation: 
Refer to the bug-reporting instructions at https://www.cgal.org/bug_report.html
libc++abi.dylib: terminating with uncaught exception of type CGAL::Assertion_exception: CGAL ERROR: assertion violation!
Expr: _idx < data_.size()
File: /usr/local/include/CGAL/Surface_mesh/Properties.h
Line: 204
Process stopped

## Source Code

_If your issue arises by using CGAL in your own source code, please provide a minimalist example that we can compile easily to reproduce the bug. If your issue arises from using a CGAL program (demo, example, etc.), please let us know which one. Helping you solving an issue is much easier and efficient if we can reproduce it._

// Adapted from Example 3.1 at https://doc.cgal.org/latest/Surface_mesh_shortest_path/index.html#Surface_mesh_shortest_pathIntroduction

```
#include <cstdlib>
#include <iostream>
#include <fstream>
#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
#include <CGAL/Random.h>
#include <CGAL/Surface_mesh.h>
#include <CGAL/Surface_mesh_shortest_path.h>
#include <boost/lexical_cast.hpp>

typedef CGAL::Exact_predicates_inexact_constructions_kernel Kernel;
typedef CGAL::Surface_mesh<Kernel::Point_3> Triangle_mesh;
typedef CGAL::Surface_mesh_shortest_path_traits<Kernel, Triangle_mesh> Traits;
typedef CGAL::Surface_mesh_shortest_path<Traits> Surface_mesh_shortest_path;
typedef boost::graph_traits<Triangle_mesh> Graph_traits;
typedef Graph_traits::vertex_iterator vertex_iterator;

int
main(int /*argc*/, char** /*argv*/) {

  const std::string modelpath = "../../models/";
  const std::string outputpath = "./dists/";

  // std::vector<std::string> models = { "bunny", "elephant", "lion-head", "fandisk-box",
  //                                     "cheese", "fandisk"};
  std::vector<std::string> models = { "lion-head"};

  // std::string model = "bunny"; {
  for( const auto& model : models) {

    // read input mesh
    Triangle_mesh tmesh;
    std::stringstream fname_model;
    fname_model << modelpath << model << ".off";
    std::ifstream input( fname_model.str());
    if( ! input.good()) {
      std::cerr << "Failed to open '" << fname_model.str() << "'. " << "Terminating." << std::endl;
      exit(-1);
    }
    input >> tmesh;
    input.close();
    std::cout << "Finished reading '" << model << "' with "
              <<  tmesh.number_of_vertices() << "." << std::endl;

    // std::size_t vidx_source = 37; {
    for( std::size_t vidx_source = 0; vidx_source < tmesh.number_of_vertices(); ++vidx_source) {

      // construct a shortest path query object and add a source point
      Surface_mesh_shortest_path shortest_paths(tmesh);
      // todo: how to access the i-th vertex more elegantly
      {
        std::size_t i = 0;
        vertex_iterator vit, vit_end;
        for ( boost::tie(vit, vit_end) = vertices(tmesh); vit != vit_end; ++vit, ++i) {
          if( vidx_source == i) {
            shortest_paths.add_source_point( *vit);
            break;
          }
        }
      }

      // For all vertices in the tmesh, compute the distance

      // auto& output = std::cout;
      std::vector<double> dists;

      vertex_iterator vit, vit_end;
      for ( boost::tie(vit, vit_end) = vertices(tmesh); vit != vit_end; ++vit) {

        std::vector<Traits::Point_3> points;
        shortest_paths.shortest_path_points_to_source_points(*vit, std::back_inserter(points));

        // compute length
        double len = 0.;
        for( std::size_t i = 1; i < points.size(); ++i) {
          len += std::sqrt( (points[i] - points[i-1]).squared_length());
        }
        dists.push_back( len);

        // // print the points
        // output << points.size() << " ";
        // for (std::size_t i = 0; i < points.size(); ++i) {
        //   output << " " << points[i];
        // }
        // output << std::endl;
      }

      // write distances to file
      std::stringstream fname;
      fname << outputpath << model << "_dists_src_" << vidx_source << ".dat";
      std::ofstream fs( fname.str(), std::ios::binary);
      if( ! fs.good()) {
        std::cerr << "Failed to open '" << fname.str() << "'. " << "Terminating." << std::endl;
        exit(-1);
      }
      fs.write( (char*) dists.data(), sizeof(double) * dists.size());
      fs.close();

      std::cout << "Finished vertex " << vidx_source << "." << std::endl;
    }
  }

  return 0;
}
```

## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits): MacOS 10.14.6 (reproduced on up-to-date Ubuntu)
* Compiler: Apple LLVM version 10.0.1 (clang-1001.0.46.4) (reproduced with gcc)
* Release or debug mode: both
* Specific flags used (if any): c++ -W -Wall -std=c++14 -lcgal -lgmp -lboost_system -lboost_thread main.cpp -o main
* CGAL version: Latest from brew
* Boost version: Latest from brew
* Other libraries versions if used (Eigen, TBB, etc.):

**Comment by MaelRL** [comment]
Do you have any way of testing the master branch of CGAL? This is likely due to your mesh having a border and this configuration had a bug which was fixed recently (https://github.com/CGAL/cgal/pull/4288); the fix is in master and in the 4.14 branch and also in the upcoming 5.0 version.

**Comment by sloriot** [comment]
Duplicate of #3993
